packages/react-server/src/ReactServerStreamConfigBrowser.js

```diff
@@ -21,87 +21,166 @@ export function flushBuffered(destination: Destination) {
-      destination.enqueue(currentView);
-      bytesToWrite = bytesToWrite.subarray(allowableBytes);
-    }
-    currentView = new Uint8Array(VIEW_SIZE);
-    writtenBytes = 0;
+    // in practice allowableBytes is never zero because eagerly flush at the
+    // end of writeChunk if we have exhausted the view
+    ((currentView: any): Uint8Array).set(
+      bytesToWrite.subarray(0, allowableBytes),
+      writtenBytes,
+    );
+    writtenBytes += allowableBytes;
+    bytesToWrite = bytesToWrite.subarray(allowableBytes);
+    flushCurrentView(destination, false);
+  }
+
+  // remaining bytesToWrite may still be too large. this can happen if the original currentView
+  // was byob and was larger than VIEW_SIZE but there wasn't enough room to reduce bytesToWrite
+  // below VIEW_SIZE
+  if (bytesToWrite.byteLength > VIEW_SIZE) {
+    // we flush here for safety but writtenBytes should be zero already because
+    // we just finished the branch above where flush was called in tail position
+    flushCurrentView(destination, false);
+    // enqueue bytesToWrite
+    destination.enqueue(bytesToWrite);
+    return;
+  } else {
+    // copy bytes to current view
+    ((currentView: any): Uint8Array).set(bytesToWrite, writtenBytes);
+    writtenBytes += bytesToWrite.byteLength;
+  }
+
+  // eagerly flush if we have fully filled the currentView
+  const remainingBytes =
+    ((currentView: any): Uint8Array).byteLength - writtenBytes;
+  if (remainingBytes === 0) {
+    flushCurrentView(destination, false);
   }
-  ((currentView: any): Uint8Array).set(bytesToWrite, writtenBytes);
-  writtenBytes += bytesToWrite.length;
 }
 
 export function writeChunkAndReturn(
   destination: Destination,
   chunk: PrecomputedChunk | Chunk,
 ): boolean {
   writeChunk(destination, chunk);
-  // in web streams there is no backpressure so we can alwas write more
+  // allow yielding if the current view is mostly full
+  const length = ((currentView: any): Uint8Array).byteLength;
+  if ((length - writtenBytes) << 2 < length) {
+    // The view is mostly full, we can yield and let the view flush
+    // it is possible that if required writes to complete a boundary overflow to
+    // a new view then we will start to return true again and not end up yielding
+    return false;
+  }
   return true;
 }
 
 export function completeWriting(destination: Destination) {
-  if (currentView && writtenBytes > 0) {
-    destination.enqueue(new Uint8Array(currentView.buffer, 0, writtenBytes));
-    currentView = null;
-    writtenBytes = 0;
-  }
+  flushCurrentView(destination, true);
+  currentView = null;
 }
 
 export function close(destination: Destination) {
   destination.close();
+  if (destination.byobRequest && destination.byobRequest.view !== null) {
+    // based on how the streams spec is written it is possible for a stream
+    // to be closed but for a curernt byobRequest to still be active.
+    // when a stream is closed you can respond to the request with 0 bytes written
```

Comment:
Did you confirm that we don't have to await the close? I tried to read the spec and I wasn't sure about the "closing" state.

Reply:
upgrade to 3.2 resolved the main way I was getting myself into this problem and assuming cloudflare is spec compliant I believe the way we are using it is ok and not possible to error but I'm going to ask about this because I do think the spec allows for the close to not proceed immediately. The main issue is I don't see a way to ascertain whether the stream is closed or just closeRequested and you cannot await the close call.

My hope is that if the stream won't close it will always have a null byobRequest (e.g. it must have been responded to already or an different enqueue invalidated it) and that the only times it will be non null are times the stream can close synchronously. I have no good reason to believe that this is in fact true so I'll ask about it on the polyfill project and see if I am missing a means to only call respond in this way when the stream is in fact closed

Reply:
I might have found a spec bug https://github.com/MattiasBuelens/web-streams-polyfill/issues/111